core/src/services/onedrive/backend.rs

```diff
@@ -64,26 +64,33 @@ impl Access for OnedriveBackend {
         }
 
         let response = self.core.onedrive_create_dir(path).await?;
-
-        let status = response.status();
-        match status {
+        match response.status() {
             StatusCode::CREATED | StatusCode::OK => Ok(RpCreateDir::default()),
             _ => Err(parse_error(response)),
         }
     }
 
-    async fn stat(&self, path: &str, _args: OpStat) -> Result<RpStat> {
-        let path = build_rooted_abs_path(&self.core.root, path);
-        let meta = self.core.onedrive_stat(path.as_str()).await?;
+    async fn stat(&self, path: &str, args: OpStat) -> Result<RpStat> {
+        let mut meta = self.core.onedrive_stat(path, args.if_none_match()).await?;
+
+        if let Some(version) = args.version() {
+            let versions = self.core.onedrive_list_versions(path).await?;
+            for item_version in versions {
+                if item_version.id == version {
+                    meta.set_version(version);
+                    return Ok(RpStat::new(meta));
+                }
+            }
+
+            return Err(Error::new(ErrorKind::NotFound, "item version not found"));
```

Comment:
This is a big sad regarding this API. 

I can imagine that `list_with(path).versions(true)` and `stat_with(item).versions(abc)` could be very slow.

---

I noticed that we're using `id` as the version ID. How about supporting `onedrive_stat_by_id`?

OneDrive supports two call conventions:

```shell
GET /me/drive/items/{item-id}
GET /me/drive/root:/{item-path}
```

If we already have the `id`, we can use it to make a direct call.

Reply:
The item ID is not the same as the version ID, unfortunately. Using versions is going to generate lots of API calls indeed. I agree with you this implementation could be better.

Reply:
> The item ID is not the same as the version ID

Sorry I didn't take deep look over the onedrive API. The `id` is not unique across the entire onedrive?

Reply:
Oh, onedrive...

Reply:
> The version ID belongs to a file.

Ok, I got it. So version ID can't be used as `item-id`.

> list_with(path).versions(true): no API support except N+1 HTTP calls

Seems it's the best we can do by given APIs.

> stat_with(item).versions(abc): one API call

Happy to know.

Reply:
Indeed it's a bit disappointing. Maybe we will find the list with version support from OneDrive a few years later.